sql/analyzer/pushdown.go

```diff
@@ -801,139 +803,108 @@ func getIndexesByTable(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scop
 	return indexes, nil
 }
 
+// replacePkSort applies an IndexAccess when there is an `OrderBy` over a prefix of any `PrimaryKey`s
 func replacePkSort(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
 	return transform.NodeWithCtx(n, nil, func(tc transform.Context) (sql.Node, transform.TreeIdentity, error) {
-		n := tc.Node
-
-		// Find order by nodes
-		s, ok := n.(*plan.Sort)
+		s, ok := tc.Node.(*plan.Sort)
 		if !ok {
-			return n, transform.SameTree, nil
+			return tc.Node, transform.SameTree, nil
 		}
-
-		// Must be sorting by ascending
-		for _, field := range s.SortFields {
-			if field.Order != sql.Ascending {
-				return n, transform.SameTree, nil
-			}
+		tableAliases, err := getTableAliases(s, scope)
+		if err != nil {
+			return tc.Node, transform.SameTree, nil
 		}
-
-		// Check for any alias projections
-		var rs *plan.ResolvedTable
-		aliasMap := make(map[string]string)
-		pj, ok := s.UnaryNode.Child.(*plan.Project)
-		var decoratingParent sql.Node
-		if ok {
-			n := pj.Child
-			if rs, ok = n.(*plan.ResolvedTable); !ok {
-				return s, transform.SameTree, nil
-			}
-			// Extract aliases
-			for _, expr := range pj.Expressions() {
-				if alias, ok := expr.(*expression.Alias); ok {
-					aliasMap[alias.Name()] = alias.UnaryExpression.Child.String()
-				}
-			}
-		} else {
-			n := s.Child
-			// Otherwise, sorts immediate child must be ResolvedTable
-			if rs, ok = n.(*plan.ResolvedTable); !ok {
-				return s, transform.SameTree, nil
+		sfExprs := normalizeExpressions(tableAliases, s.SortFields.ToExpressions()...)
+		skipSel := func(tc transform.Context) bool {
+			switch tc.Node.(type) {
+			// TODO: possible that we should fix these
+			case *plan.JoinNode, *plan.Distinct:
+				return false
+			default:
+				return true
 			}
 		}
-
-		// Extract primary key columns from index to maintain order
-		table := rs.Table
-		if w, ok := table.(sql.TableWrapper); ok {
-			table = w.Underlying()
-		}
-		idxTbl, ok := table.(sql.IndexAddressableTable)
-		if !ok {
-			return s, transform.SameTree, nil
-		}
-		idxs, err := idxTbl.GetIndexes(ctx)
-		if err != nil {
-			return nil, transform.SameTree, err
-		}
-
-		// Extract primary index
-		var pkIndex sql.Index
-		for _, idx := range idxs {
-			oi, ok := idx.(sql.OrderedIndex)
-			if !ok || oi.Order() != sql.IndexOrderAsc {
-				continue
+		sfAliases := aliasedExpressionsInNode(s)
+		newN, same, err := transform.NodeWithCtx(s, skipSel, func(tc transform.Context) (sql.Node, transform.TreeIdentity, error) {
```

Comment:
context-based transforms are more expensive than regular transforms, we should maybe default to `transform.Node` unless we need the parent or child index metadata

Reply:
I swtiched to using `transform.NodeWithCtx` because his optimization should be skipped for chlidren of `join`s (unless in a subquery). `transform.Node` is a bottom up transformation, while `transform,.NodeWithCtx` calls the selector beforehand.